Applications/Spreadsheet/HelpWindow.cpp

```diff
@@ -124,19 +124,26 @@ HelpWindow::HelpWindow(GUI::Window* parent)
                 return;
             }
 
-            auto dialog = GUI::Window::construct(this);
-            dialog->resize(size());
-            dialog->set_icon(icon());
-            dialog->set_title(String::formatted("Spreadsheet Help - Example {} for {}", name, entry));
-            auto& widget = dialog->set_main_widget<SpreadsheetWidget>(NonnullRefPtrVector<Sheet> {}, false);
+            auto window = GUI::Window::construct(this);
+            m_all_example_windows.set(window.ptr(), window);
+            window->resize(size());
+            window->set_icon(icon());
+            window->set_title(String::formatted("Spreadsheet Help - Example {} for {}", name, entry));
+            auto& widget = window->set_main_widget<SpreadsheetWidget>(NonnullRefPtrVector<Sheet> {}, false);
             auto sheet = Sheet::from_json(value.as_object(), widget.workbook());
             if (!sheet) {
                 GUI::MessageBox::show_error(this, String::formatted("Corrupted example '{}' in '{}'", name, url.path()));
                 return;
             }
 
             widget.add_sheet(sheet.release_nonnull());
-            dialog->show();
+            window->show();
+            window->on_close_request = [this, key = window.ptr()] {
+                auto window = m_all_example_windows.get(key).value();
+                m_all_example_windows.remove(key);
+                const_cast<GUI::Window*>(window)->remove_from_parent();
```

Comment:
I feel like this is a bit hacky, not sure why `GUI::Window::construct(parent)` has to create a strong reference to the new object in the parent, but this is what I came up with :shrug:

Reply:
Why do you need `m_all_example_windows` at all?

Reply:
Another option is to copy `window` into this lambda to be able to call `remove_from_parent()` on it. but that's a reference cycle and I'd rather not introduce any of those

Reply:
Am I missing an obvious solution here?

Reply:
What I've done elswhere is capturing a raw pointer and using `on_close`:

```cpp
    properties->on_close = [properties = properties.ptr()] {
        properties->remove_from_parent();
    };
```

Reply:
It's not the most beautiful thing, but meh. Another alternative might be to have `on_close` include  a `Window&` parameter.

Reply:
That's what I was trying to avoid, but if it's there already, I'll change this to match.

Perhaps a better suggestion, why not just remove the window from its parent when it's "closed"?

Reply:
I've specifically tried to avoid those kinds of magical ownership behaviors in LibCore/LibGUI. Some GUI toolkits work that way and it makes lifetime very confusing.

Reply:
is it really magical though? I fully expect a window to be deleted when
- I have no references to it
- it's closed

if anything, I think the parent holding on to it is more surprising than not.

Reply:
Hm, maybe you're right. It does seem like a fairly intuitive behavior. Let's do it.

Reply:
I tried this, but it doesn't look like we can reliably detect "only referenced by parent object and being closed".
a window with no external refs has a refcount of 3 in its `close()` (when invoked through the close button)
The same could also happen if it's called from outside the event handlers, where it would have refcount == 3.
Unconditionally removing from the parent child list isn't really an option either, it would also be really surprising.

I'll just leave it be the way it is for now